src/main/java/org/fungover/thunder/PackageReader.java

```diff
@@ -0,0 +1,46 @@
+package org.fungover.thunder;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.util.HashMap;
+import java.util.Map;
+
+public class PackageReader {
+    private final Map<InetAddress, Boolean> connectPackageSent = new HashMap<>();
+
+    public boolean isValidConnection(Socket socket) throws IOException {
+        InetAddress client = socket.getInetAddress();
+
+        if (connectPackageSent.containsKey(client) && Boolean.TRUE.equals(connectPackageSent.get(client))) {
+            return false;
+        }
+
+        socket.setSoTimeout(30);
+        InputStream inputStream = socket.getInputStream();
+        OutputStream outputStream = socket.getOutputStream();
+        byte[] buffer = new byte[1024];
+        int bytesRead = inputStream.read(buffer);
+
+        if (isConnectPackage(bytesRead, buffer)) {
+            System.out.println("Received MQTT CONNECT message from client");
+            connectPackageSent.put(socket.getInetAddress(),true);
+            sendConnackToClient(outputStream);
+            System.out.println("Sent MQTT CONNACK message to client");
+        }
+
+        return false;
```

Comment:
This needs to return true, if it returns false the socket will close and disconnect the client.

Reply:
Correct, I did forget that. Thank you! 😄